README.md
🚨🚨🚨

**This library is no longer actively maintained.** The Authy API has been replaced with the [Twilio Verify API](https://www.twilio.com/docs/verify). Twilio will support the Authy API through November 1, 2022 for SMS/Voice. After this date, we’ll start to deprecate the service for SMS/Voice. Any requests sent to the API after May 1, 2023, will automatically receive an error.  Push and TOTP will continue to be supported through July 2023.

[Learn more about migrating from Authy to Verify.](https://www.twilio.com/blog/migrate-authy-to-verify)

Please visit the Twilio Docs for:
* [Verify + Ruby (Rails) quickstart](https://www.twilio.com/docs/verify/quickstarts/ruby-rails)
* [Twilio Ruby helper library](https://www.twilio.com/docs/libraries/ruby)
* [Verify API reference](https://www.twilio.com/docs/verify/api)
* **Coming soon**: Look out for a new Devise plugin to use Twilio Verify with Devise

Please direct any questions to [Twilio Support](https://support.twilio.com/hc/en-us). Thank you!

🚨🚨🚨

---

# Authy Devise [![Build Status](https://github.com/twilio/authy-devise/workflows/build/badge.svg)](https://github.com/twilio/authy-devise/actions)

This is a [Devise](https://github.com/plataformatec/devise) extension to add [Two-Factor Authentication with Authy](https://www.twilio.com/docs/authy) to your Rails application.

* [Pre-requisites](#pre-requisites)
* [Demo](#demo)
* [Getting started](#getting-started)
  * [Configuring Models](#configuring-models)
    * [With the generator](#with-the-generator)
    * [Manually](#manually)
    * [Final steps](#final-steps)
* [Custom Views](#custom-views)
  * [Request a phone call](#request-a-phone-call)
* [Custom Redirect Paths (eg. using modules)](#custom-redirect-paths-eg-using-modules)
* [I18n](#i18n)
* [Session variables](#session-variables)
* [OneTouch support](#onetouch-support)
* [Generic authenticator token support](#generic-authenticator-token-support)
* [Rails 5 CSRF protection](#rails-5-csrf-protection)
* [Running Tests](#running-tests)
* [Notice: Twilio Authy API’s Sandbox feature will stop working on Sep 30, 2021](#notice-twilio-authy-apis-sandbox-feature-will-stop-working-on-sep-30-2021)
* [Copyright](#copyright)

## Pre-requisites

To use the Authy API you will need a Twilio Account, [sign up for a free Twilio account here](https://www.twilio.com/try-twilio).

Create an [Authy Application in the Twilio console](https://www.twilio.com/console/authy/applications) and take note of the API key.

## Demo

See [this repo for a full demo of using `authy-devise`](https://github.com/twilio/authy-devise-demo).

## Getting started

First get your Authy API key from [the Twilio console](https://www.twilio.com/console/authy/applications). We recommend you store your API key as an environment variable.

```bash
$ export AUTHY_API_KEY=YOUR_AUTHY_API_KEY
```

Next add the gem to your Gemfile:

```ruby
gem 'devise'
gem 'devise-authy'
```

And then run `bundle install`

Add `Devise Authy` to your App:

    rails g devise_authy:install

    --haml: Generate the views in Haml
    --sass: Generate the stylesheets in Sass

### Configuring Models

You can add devise_authy to your user model in two ways.

#### With the generator

Run the following command:

```bash
rails g devise_authy [MODEL_NAME]
```

To support account locking (recommended), you must add `:authy_lockable` to the `devise :authy_authenticatable, ...` configuration in your model as this is not yet supported by the generator.

#### Manually

Add `:authy_authenticatable` and `:authy_lockable` to the `devise` options in your Devise user model:

```ruby
devise :authy_authenticatable, :authy_lockable, :database_authenticatable, :lockable
```

(Note, `:authy_lockable` is optional but recommended. It should be used with Devise's own `:lockable` module).

Also add a new migration. For example, if you are adding to the `User` model, use this migration:

```ruby
class DeviseAuthyAddToUsers < ActiveRecord::Migration[6.0]
  def self.up
    change_table :users do |t|
      t.string    :authy_id
      t.datetime  :last_sign_in_with_authy
      t.boolean   :authy_enabled, :default => false
    end

    add_index :users, :authy_id
  end

  def self.down
    change_table :users do |t|
      t.remove :authy_id, :last_sign_in_with_authy, :authy_enabled
    end
  end
end
```

#### Final steps

For either method above, run the migrations:

```bash
rake db:migrate
```

**[Optional]** Update the default routes to point to something like:

```ruby
devise_for :users, :path_names => {
	:verify_authy => "/verify-token",
	:enable_authy => "/enable-two-factor",
	:verify_authy_installation => "/verify-installation",
	:authy_onetouch_status => "/onetouch-status"
}
```

Now whenever a user wants to enable two-factor authentication they can go to:

    http://your-app/users/enable-two-factor

And when the user logs in they will be redirected to:

    http://your-app/users/verify-token

## Custom Views

If you want to customise your views, you can modify the files that are located at:

    app/views/devise/devise_authy/enable_authy.html.erb
    app/views/devise/devise_authy/verify_authy.html.erb
    app/views/devise/devise_authy/verify_authy_installation.html.erb

### Request a phone call

The default views come with a button to force a request for an SMS message. You can also add a button that will request a phone call instead. Simply add the helper method to your view:

    <%= authy_request_phone_call_link %>

## Custom Redirect Paths (eg. using modules)

If you want to customise the redirects you can override them within your own controller like this:

```ruby
class MyCustomModule::DeviseAuthyController < Devise::DeviseAuthyController

  protected
    def after_authy_enabled_path_for(resource)
      my_own_path
    end

    def after_authy_verified_path_for(resource)
      my_own_path
    end

    def after_authy_disabled_path_for(resource)
      my_own_path
    end

    def invalid_resource_path
      my_own_path
    end
end
```

And tell the router to use this controller

```ruby
devise_for :users, controllers: {devise_authy: 'my_custom_module/devise_authy'}
```

## I18n

The install generator also copies a `Devise Authy` i18n file which you can find at:

    config/locales/devise.authy.en.yml

## Session variables

If you want to know if the user is signed in using Two-Factor authentication,
you can use the following session variable:

```ruby
session["#{resource_name}_authy_token_checked"]

# Eg.
session["user_authy_token_checked"]
```

## OneTouch support

To enable [Authy push authentication](https://www.twilio.com/authy/features/push), you need to modify the Devise config file `config/initializers/devise.rb` and add configuration:

```
config.authy_enable_onetouch = true
```

## Generic authenticator token support

Authy supports other authenticator apps by providing a QR code that your users can scan.

> **To use this feature, you need to enable it in your [Twilio Console](https://www.twilio.com/console/authy/applications)**

Once you have enabled generic authenticator tokens, you can enable this in devise-authy by modifying the Devise config file `config/initializers/devise.rb` and adding the configuration:

```
config.authy_enable_qr_code = true
```

This will display a QR code on the verification screen (you still need to take a user's phone number and country code). If you have implemented your own views, the QR code URL is available on the verification page as `@authy_qr_code`.

## Rails 5 CSRF protection

In Rails 5 `protect_from_forgery` is no longer prepended to the `before_action` chain. If you call `authenticate_user` before `protect_from_forgery` your request will result in a "Can't verify CSRF token authenticity" error.

To remedy this, add `prepend: true` to your `protect_from_forgery` call, like in this example from the [Authy Devise demo app](https://github.com/twilio/authy-devise-demo):

```ruby
class ApplicationController < ActionController::Base
  protect_from_forgery with: :exception, prepend: true
end
```

## Running Tests

Run the following command:

```bash
$ bundle exec rspec
```

## Notice: Twilio Authy API’s Sandbox feature will stop working on Sep 30, 2021
Twilio is discontinuing the Authy API’s Sandbox, a feature that allows customers to run continuous integration tests against a mock Authy API for free. The Sandbox is no longer being maintained, so we will be taking the final deprecation step of shutting it down on September 30, 2021. The rest of the Authy API product will continue working as-is.

This repo previously used the sandbox API as part of the test suite, but that has been since removed.

You will only be affected if you are using the sandbox API in your own application or test suite.

For more information please read this article on [how we are discontinuing the Twilio Authy sandbox API](https://support.authy.com/hc/en-us/articles/1260803396889-Notice-Twilio-Authy-API-s-Sandbox-feature-will-stop-working-on-Sep-30-2021).

## Copyright

Copyright (c) 2012-2021 Authy Inc. See LICENSE.txt for further details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.3.1] - 2022-05-30 Final release

### Changed

- Added deprecation notices to README and Gemspec

## [2.3.0] - 2021-01-07

### Fixed

- Fixes calls to `I18n.t` with keyword arguments to support Ruby 3.0
- Replaces Travis CI with GitHub Actions
- Updates webmock development dependency
- Removes sdoc from Gemfile

## [2.2.1] - 2020-10-13

### Fixed

- If the app offers a QR code scan and user fails to verify authy installation, the QR code wasn't shown again. Fixed in (#149)

## [2.2.0] - 2020-06-04

### Fixed

- Don't delete user in Authy if another user has the same authy_id (#144)

## [2.1.0] - 2020-05-05

### Added

- Support for generic authenticator tokens (#141)

### Fixed

- Can remember device when enabling 2FA for the first time (#139)

## [2.0.0] - 2020-04-28

Releasing this as version 2 because there is a significant change in dependencies. Minimum version of Rails is now 5 and of Devise is now 4. Otherwise the gem should work as before.

### Added

- HTTP Only flag to remember_device cookie (#116 thanks @agronv)
- Remembers device when user logs in with One Touch (#128 thanks @cplopez4)
- Autocomplete attributes for HTML form (#130)

### Changed

- Mocked API calls in test suite (#123)
- Full test suite refactor (#124)
- Increased required version for Devise and Rails (#125)
- Stopped calling `signed_in?` before it is needed (#126)

### Fixes

- Remembers user correctly when logging in with One Touch (#129)

## [1.11.1] - 2019-02-02

### Fixed

- Using the version before loading it broke everything. :facepalm:

## [1.11.0] - 2019-02-01

### Fixed

- Corrects for label in verify_authy view (#103 thanks @mstruebing)
- Corrects heading in verify_authy view (#104 thanks @mstruebing)

### Changed

- Allows you to define paths for request_sms and request_phone_call (#108 thanks @dedene)

### Added

- Now sets a distinct user agent through the Authy gem (#110)

## [1.10.0] - 2018-09-26

### Changed

- Moves OneTouch approval request copy to locale file.

### Removed

- Demo app now lives in its own repo

## [1.9.0] - 2018-09-04

### Fixed

- Generated migration now includes version number for Rails 5

### Changed

- Removes Jeweler in favour of administering the gemspec by hand
- Removes demo app files from gem package

## [1.8.3] - 2018-07-05

### Fixed

- Fixes Ruby interpolation in HAML for onetouch (thanks @muan)
- Records Authy authentication after install verification (thanks @nukturnal)
- Forgets remember device cookie when disabling Authy (thanks @senekis)

### Changed

- Updated testing Rubies in CI

## Older releases

**_The following releases happened before the changelog was started. Some history will be added for clarity._**

## [1.8.2] - 2017-12-22

## [1.8.1] - 2016-12-06

## [1.8.0] - 2016-10-25

## [1.7.0] - 2015-12-22

## [1.6.0] - 2015-01-07

## [1.5.3] - 2014-06-11

## [1.5.2] - 2014-06-11

## [1.5.1] - 2014-04-24

## [1.5.0] - 2014-01-07

## [1.4.0] - 2013-12-17

## [1.3.0] - 2013-11-16

## [1.2.2] - 2013-09-04

## [1.2.1] - 2013-04-22

## [1.2.0] - 2013-04-22 [YANKED]

## [1.0.0] - 2013-04-10


app/assets/javascripts/form.authy.js
var Authy = {};

// Fix Internet Explorer by defining getElementsByClassName
if(document.getElementsByClassName == null) {
    document.getElementsByClassName = function(className, parentElement) {
        var children = (parentElement || document.body).getElementsByTagName('*');
        var elements = [], child;
        for (var i = 0, length = children.length; i < length; i++) {
            child = children[i];
            if ((' '+child.className+' ').indexOf(' '+className+' ') != -1){
                elements.push(child);
            }
        }
        return elements;
    };

    HTMLDivElement.prototype.getElementsByClassName = function(className) {
        return document.getElementsByClassName(className, this);
    }
};

Authy.UI = function() {
    // Attributes
    var self = this;
    var tooltipTitle = "Authy Help";
    var tooltipMessage = 'This token is only needed if you setup Authy to protect your account. Learn more <a href="https://www.authy.com/help">here</a>';

    var countriesList = [
        {"country":"United States of America","code":"1"},{"country":"Canada","code":"1"},{"country":"Russia","code":"7"},{"country":"Kazakhstan","code":"7"},
        {"country":"Egypt","code":"20"},{"country":"South Africa","code":"27"},{"country":"Greece","code":"30"},{"country":"Netherlands","code":"31"},
        {"country":"Belgium","code":"32"},{"country":"France","code":"33"},{"country":"Spain","code":"34"},{"country":"Hungary","code":"36"},
        {"country":"Italy","code":"39"},{"country":"Romania","code":"40"},{"country":"Switzerland","code":"41"},{"country":"Austria","code":"43"},
        {"country":"United Kingdom","code":"44"},{"country":"Guernsey","code":"44"},{"country":"Isle of Man","code":"44"},{"country":"Jersey","code":"44"},
        {"country":"Denmark","code":"45"},{"country":"Sweden","code":"46"},{"country":"Norway","code":"47"},{"country":"Poland","code":"48"},
        {"country":"Germany","code":"49"},{"country":"Peru","code":"51"},{"country":"Mexico","code":"52"},{"country":"Cuba","code":"53"},
        {"country":"Argentina","code":"54"},{"country":"Brazil","code":"55"},{"country":"Chile","code":"56"},{"country":"Colombia","code":"57"},
        {"country":"Venezuela","code":"58"},{"country":"Malaysia","code":"60"},{"country":"Australia","code":"61"},{"country":"Indonesia","code":"62"},
        {"country":"Philippines","code":"63"},{"country":"New Zealand","code":"64"},{"country":"Singapore","code":"65"},{"country":"Thailand","code":"66"},
        {"country":"Japan","code":"81"},{"country":"Korea (South)","code":"82"},{"country":"Vietnam","code":"84"},{"country":"China","code":"86"},
        {"country":"Turkey","code":"90"},{"country":"India","code":"91"},{"country":"Pakistan","code":"92"},{"country":"Afghanistan","code":"93"},
        {"country":"Sri Lanka","code":"94"},{"country":"Myanmar","code":"95"},{"country":"Iran","code":"98"},{"country":"Morocco","code":"212"},
        {"country":"Algeria","code":"213"},{"country":"Tunisia","code":"216"},{"country":"Libya","code":"218"},{"country":"Gambia","code":"220"},
        {"country":"Senegal","code":"221"},{"country":"Mauritania","code":"222"},{"country":"Mali Republic","code":"223"},{"country":"Guinea","code":"224"},
        {"country":"Ivory Coast","code":"225"},{"country":"Burkina Faso","code":"226"},{"country":"Niger","code":"227"},{"country":"Togo","code":"228"},
        {"country":"Benin","code":"229"},{"country":"Mauritius","code":"230"},{"country":"Liberia","code":"231"},{"country":"Sierra Leone","code":"232"},
        {"country":"Ghana","code":"233"},{"country":"Nigeria","code":"234"},{"country":"Chad","code":"235"},{"country":"Central African Republic","code":"236"},
        {"country":"Cameroon","code":"237"},{"country":"Cape Verde Islands","code":"238"},{"country":"Sao Tome and Principe","code":"239"},
        {"country":"Gabon","code":"241"},{"country":"Congo, Democratic Republ","code":"243"},{"country":"Angola","code":"244"},
        {"country":"Guinea-Bissau","code":"245"},{"country":"Seychelles","code":"248"},{"country":"Sudan","code":"249"},{"country":"Rwanda","code":"250"},
        {"country":"Ethiopia","code":"251"},{"country":"Somalia","code":"252"},{"country":"Djibouti","code":"253"},{"country":"Kenya","code":"254"},
        {"country":"Tanzania","code":"255"},{"country":"Uganda","code":"256"},{"country":"Burundi","code":"257"},{"country":"Mozambique","code":"258"},
        {"country":"Zambia","code":"260"},{"country":"Madagascar","code":"261"},{"country":"Reunion","code":"262"},{"country":"Zimbabwe","code":"263"},
        {"country":"Namibia","code":"264"},{"country":"Malawi","code":"265"},{"country":"Lesotho","code":"266"},{"country":"Botswana","code":"267"},
        {"country":"Swaziland","code":"268"},{"country":"Mayotte Island","code":"269"},{"country":"Aruba","code":"297"},{"country":"Faroe Islands","code":"298"},
        {"country":"Greenland","code":"299"},{"country":"Gibraltar","code":"350"},{"country":"Portugal","code":"351"},{"country":"Luxembourg","code":"352"},
        {"country":"Ireland","code":"353"},{"country":"Iceland","code":"354"},{"country":"Albania","code":"355"},{"country":"Malta","code":"356"},
        {"country":"Cyprus","code":"357"},{"country":"Finland","code":"358"},{"country":"Bulgaria","code":"359"},{"country":"Lithuania","code":"370"},
        {"country":"Latvia","code":"371"},{"country":"Estonia","code":"372"},{"country":"Moldova","code":"373"},{"country":"Armenia","code":"374"},
        {"country":"Belarus","code":"375"},{"country":"Andorra","code":"376"},{"country":"Monaco","code":"377"},{"country":"San Marino","code":"378"},
        {"country":"Ukraine","code":"380"},{"country":"Serbia","code":"381"},{"country":"Montenegro","code":"382"},{"country":"Croatia","code":"385"},
        {"country":"Slovenia","code":"386"},{"country":"Bosnia-Herzegovina","code":"387"},{"country":"Macedonia","code":"389"},
        {"country":"Czech Republic","code":"420"},{"country":"Slovakia","code":"421"},{"country":"Liechtenstein","code":"423"},
        {"country":"Falkland Islands","code":"500"},{"country":"Belize","code":"501"},{"country":"Guatemala","code":"502"},{"country":"El Salvador","code":"503"},
        {"country":"Honduras","code":"504"},{"country":"Nicaragua","code":"505"},{"country":"Costa Rica","code":"506"},{"country":"Panama","code":"507"},
        {"country":"Haiti","code":"509"},{"country":"Guadeloupe","code":"590"},{"country":"Bolivia","code":"591"},{"country":"Guyana","code":"592"},
        {"country":"Ecuador","code":"593"},{"country":"French Guiana","code":"594"},{"country":"Paraguay","code":"595"},{"country":"Martinique","code":"596"},
        {"country":"Suriname","code":"597"},{"country":"Uruguay","code":"598"},{"country":"Netherlands Antilles","code":"599"},{"country":"Timor-Leste","code":"670"},
        {"country":"Guam","code":"671"},{"country":"Brunei","code":"673"},{"country":"Nauru","code":"674"},{"country":"Papua New Guinea","code":"675"},
        {"country":"Tonga","code":"676"},{"country":"Solomon Islands","code":"677"},{"country":"Vanuatu","code":"678"},{"country":"Fiji Islands","code":"679"},
        {"country":"Cook Islands","code":"682"},{"country":"Samoa","code":"685"},{"country":"New Caledonia","code":"687"},{"country":"French Polynesia","code":"689"},
        {"country":"Korea (North)","code":"850"},{"country":"HongKong","code":"852"},{"country":"Macau","code":"853"},{"country":"Cambodia","code":"855"},
        {"country":"Laos","code":"856"},{"country":"Bangladesh","code":"880"},{"country":"International","code":"882"},{"country":"Taiwan","code":"886"},
        {"country":"Maldives","code":"960"},{"country":"Lebanon","code":"961"},{"country":"Jordan","code":"962"},{"country":"Syria","code":"963"},
        {"country":"Iraq","code":"964"},{"country":"Kuwait","code":"965"},{"country":"Saudi Arabia","code":"966"},{"country":"Yemen","code":"967"},
        {"country":"Oman","code":"968"},{"country":"Palestine","code":"970"},{"country":"United Arab Emirates","code":"971"},{"country":"Israel","code":"972"},
        {"country":"Bahrain","code":"973"},{"country":"Qatar","code":"974"},{"country":"Bhutan","code":"975"},{"country":"Mongolia","code":"976"},
        {"country":"Nepal","code":"977"},{"country":"Tajikistan","code":"992"},{"country":"Turkmenistan","code":"993"},{"country":"Azerbaijan","code":"994"},
        {"country":"Georgia","code":"995"},{"country":"Kyrgyzstan","code":"996"},{"country":"Uzbekistan","code":"998"},{"country":"Bahamas","code":"1242"},
        {"country":"Barbados","code":"1246"},{"country":"Anguilla","code":"1264"},{"country":"Antigua and Barbuda","code":"1268"},
        {"country":"Virgin Islands, British","code":"1284"},{"country":"Cayman Islands","code":"1345"},{"country":"Bermuda","code":"1441"},
        {"country":"Grenada","code":"1473"},{"country":"Turks and Caicos Islands","code":"1649"},{"country":"Montserrat","code":"1664"},
        {"country":"Saint Lucia","code":"1758"},{"country":"Dominica","code":"1767"},{"country":"St. Vincent and The Gren","code":"1784"},
        {"country":"Puerto Rico","code":"1787"},{"country":"Dominican Republic","code":"1809"},{"country":"Dominican Republic2","code":"1829"},
        {"country":"Dominican Republic3","code":"1849"},{"country":"Trinidad and Tobago","code":"1868"},{"country":"Saint Kitts and Nevis","code":"1869"},
        {"country":"Jamaica","code":"1876"},{"country":"Congo","code":"2420"}
    ];
    var countries = [];

    // Private Members
    var setupCellphoneValidation = function() {
        var cellPhone = document.getElementById("authy-cellphone");
        if(!cellPhone) return;

        cellPhone.onblur = function(){
          if(cellPhone.value != '' && cellPhone.value.match(/^([0-9][0-9][0-9])\W*([0-9][0-9]{2})\W*([0-9]{3,5})$/)){
            cellPhone.style.backgroundColor = "white";
          } else {
            cellPhone.style.backgroundColor = "#F2DEDE";
          }
        }
    };

    var setupAuthyToken = function() {
        var token = document.getElementById("authy-token");
        if(!token) return;

        token.onblur = function() {
          if(token.value != '' && token.value.match(/^\d+$/)){
            token.style.backgroundColor = "white";
          } else {
            token.style.backgroundColor = "#F2DEDE";
          }
        };
    };

    var setupTooltip = function() {
        var authy_help = document.getElementById("authy-help");
        if(!authy_help) return;

        var tooltip = document.createElement("div");
        tooltip.setAttribute("id", "authy-tooltip");

        var pos = absolutePosFor(authy_help);

        var tooltipTop = pos[0];
        var tooltipLeft = pos[1]+10;



        tooltip.setAttribute("style", "top:" + tooltipTop + "px;left:" + tooltipLeft + "px;");
        tooltip.innerHTML = '<a id="authy-tooltip-close"></a><h3 class="tooltip-title">'+tooltipTitle+'</h3><p class="tooltip-content">'+tooltipMessage+'</p>';
        document.body.appendChild(tooltip);
        document.getElementById('authy-help').onclick = function(){ document.getElementById("authy-tooltip").style.display = "block"; };
        document.getElementById('authy-tooltip-close').onclick = function(){ document.getElementById("authy-tooltip").style.display = "none"; };
    };

    var processKey40 = function(listId) {
        var caId = 'countries-autocomplete-'+listId;
        var countriesArr = document.getElementById(caId).getElementsByTagName('li');
        for(var i=0; i<countriesArr.length; i++){
            if(document.getElementById(caId).getElementsByTagName('li')[i].className == 'active'){
                document.getElementById(caId).getElementsByTagName('li')[i].className = '';
                if((i+1) == countriesArr.length){
                    document.getElementById(caId).getElementsByTagName('li')[0].className = 'active';
                } else {
                    document.getElementById(caId).getElementsByTagName('li')[i+1].className = 'active';
                }
                return false;
            }
        }
        document.getElementById(caId).getElementsByTagName('li')[0].setAttribute('class', 'active');
    };

    var processKey38 = function(listId) {
        var caId = 'countries-autocomplete-'+listId;
        var countriesArr = document.getElementById(caId).getElementsByTagName('li');
        for(var i=countriesArr.length-1; i>=0; i--){
            if(document.getElementById(caId).getElementsByTagName('li')[i].className == 'active'){
                document.getElementById(caId).getElementsByTagName('li')[i].className = '';
                if(i == 0){
                    document.getElementById(caId).getElementsByTagName('li')[countriesArr.length-1].className = 'active';
                } else {
                    document.getElementById(caId).getElementsByTagName('li')[i-1].className = 'active';
                }
                return false;
            }
        }
        document.getElementById(caId).getElementsByTagName('li')[0].setAttribute('class', 'active');
    }

    var processKey13 = function(listId) {
        var obj = document.getElementById('countries-autocomplete-'+listId).getElementsByClassName('active')[0];

        self.autocomplete(obj);
        return false;
    }

    var setupEvents = function(countriesInput, listId) {
        if(!countriesInput) return;

        countriesInput.onfocus = function() {
            document.getElementById('countries-autocomplete-'+listId).style.display = "block";
        }

        // when it loses focus hide the list
        countriesInput.onblur = function() {
            // defer this call so click events can be processed
            setTimeout(function() {
              document.getElementById('countries-autocomplete-'+listId).style.display = "none";
            }, 200);
        }

        countriesInput.onkeyup = function(event) {
            document.getElementById('countries-autocomplete-'+listId).style.display = "block";
            var keyID = window.event.keyCode; // IE

            switch(keyID) {
                case 13:
                    // enter
                    processKey13(listId);
                    return false;
                break;
                case 40:
                    // down key
                    if(processKey40(listId) == false)
                        return false;
                break;
                case 38:
                    // up key
                    if(processKey38(listId) == false)
                        return false;
                break;
            }

            self.searchItem(listId);
        };

        countriesInput.onkeypress = function(event) {
          if(window.event.keyCode == 13) {
              processKey13(listId); // Enter key
              return false;
          }
        };

        document.getElementById('countries-autocomplete-'+listId).onclick = function(e){
            if(e && e.stopPropagation) {
                e.stopPropagation();
            } else {
                e = window.event;
                e.cancelBubble = true;
            }
        }

        document.getElementById('countries-input-'+listId).onclick = function(e){
            if(e && e.stopPropagation) {
                e.stopPropagation();

                countriesInput.focus();
                countriesInput.select();
            } else {
                e = window.event;
                e.cancelBubble = true;
            }
        }

        document.onclick = function() {
            document.getElementById('countries-autocomplete-'+listId).style.display = "none";
        }


    };

    var buildItem = function(classActive, country, listId) {
        var cc = country.country.substring(0,2).toLowerCase()+country.code;
        return '<li class="' + classActive + '" onclick="Authy.UI.instance().autocomplete(this);return false;" data-list-id="'+listId+'" rel="' + country.code + '" data-name="' + country.country + '"'+ '>'+
               '<span class="aflag flag_' + cc + '"></span> '+
               ' <span>(+'+ country.code + ') ' + country.country + '</span></li>';
    }

    var absolutePosFor = function(element) {
        var absTop = 0;
        var absLeft = 0;
        while(element) {
            absTop += element.offsetTop;
            absLeft += element.offsetLeft;

            element = element.offsetParent;
        }

        return [absTop, absLeft];
    }

    var setupCountriesDropdown = function(countriesSelect, listId) {
        if(!countriesSelect) {
          return;
        }

        var countries = [];
        for( var i=0; i < countriesSelect.getElementsByTagName("option").length; i++){
            var buf = [];
            buf[0] = countriesSelect.getElementsByTagName("option")[i].value;
            buf[1] = countriesSelect.getElementsByTagName("option")[i].innerHTML;
            countries.push(buf);
        }

        countriesSelect.setAttribute("style", "display:none");
        var name = countriesSelect.getAttribute("name");
        countriesSelect.removeAttribute("name");

        var countriesAutocomplete = document.createElement("div");
        var countryCodeValue = document.createElement("input");
        countryCodeValue.setAttribute("type", "hidden");
        countryCodeValue.setAttribute("id", "country-code-"+listId);
        countryCodeValue.setAttribute("name", name);

        var classActive = '';

        var countriesAutocompleteHTML = '<ul>';

        for( var i=0; i < countriesList.length; i++){
            classActive = (i == 0) ? 'active' : '';
            countriesAutocompleteHTML += buildItem(classActive, countriesList[i], listId);
        }
        countriesAutocompleteHTML += '</ul>';
        countriesAutocomplete.innerHTML = countriesAutocompleteHTML;

        document.body.appendChild(countriesAutocomplete);

        var countriesInput = document.createElement("input");
        countriesInput.setAttribute("id", "countries-input-"+listId);
        countriesInput.setAttribute("class", "countries-input")
        countriesInput.setAttribute("type", "text");
        countriesInput.setAttribute("autocomplete", "off");

        countriesSelect.parentNode.insertBefore(countriesInput, countriesSelect);
        countriesSelect.parentNode.appendChild(countryCodeValue);

        var pos = absolutePosFor(countriesInput);

        countriesAutocomplete.setAttribute("id", "countries-autocomplete-"+listId);
        countriesAutocomplete.setAttribute("class", "countries-autocomplete");
        countriesAutocomplete.setAttribute("style", "width: "+(countriesInput.offsetWidth-5)+"px; top: " + (pos[0]+27)  + "px; left: " + pos[1]  + "px;");

        setupEvents(countriesInput, listId)
    }

    var findAndSetupCountries = function() {
        setupCountriesDropdown(document.getElementById("authy-countries"), 0);
        var countries = document.getElementsByClassName("authy-countries");

        for(var i = 0; i < countries.length; i++) {
          setupCountriesDropdown(countries[i], i+1);
        }
    };



    // Public Members
    this.init = function() {
        setupAuthyToken();
        setupTooltip();
        findAndSetupCountries();
        setupCellphoneValidation();
    }

    this.searchItem = function(listId) {
        var classActive = 'active';
        var countriesInput = document.getElementById('countries-input-'+listId);
        var str = countriesInput.value;
        var countriesAutocompleteHTML = '<ul>';
        var ok = false;

        str = str.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&"); // Escape regular expression special characters

        var reg = new RegExp("^"+str, "i");
        for( var i=0; i < countriesList.length; i++){
            var countryItem = countriesList[i];
            var countryWords = countryItem.country.toLowerCase().split(/\s+/);
            // Capture words starting with the entered pattern
            for(var cw = 0; cw < countryWords.length; cw++) {
                if(countryWords[cw].length > 2 && countryWords[cw].match(reg)) {
                    countriesAutocompleteHTML += buildItem(classActive, countryItem, listId);
                    classActive = '';
                    ok = true;
                    break;
                }
            }
        }
        countriesAutocompleteHTML += '</ul>';
        if(ok) {
            document.getElementById('countries-autocomplete-'+listId).innerHTML = countriesAutocompleteHTML;
        }
    }

    this.autocomplete = function(obj) {
        var listId = obj.getAttribute("data-list-id");
        document.getElementById('countries-input-'+listId).value = obj.getAttribute("data-name");
        document.getElementById('countries-autocomplete-'+listId).style.display = "none";

        document.getElementById('country-code-'+listId).value = obj.getAttribute('rel');
    }

    this.setTooltip = function(title, msg) {
        var tooltip = document.getElementById("authy-tooltip");
        if(!tooltip) return;

        tooltip.getElementsByClassName("tooltip-title")[0].innerHTML = title;
        tooltip.getElementsByClassName("tooltip-content")[0].innerHTML = msg;
    }
};

Authy.UI.instance = function() {
    if(!this.ui){
        this.ui = new Authy.UI();
        this.ui.init();
    }

    return this.ui;
}

window.onload = function() {
    Authy.UI.instance();
}
